pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,464 @@
+// Use warning messages to describe problems the client making the API request should correct or be aware of
+// Limit warnings to 120 characters if possible
+func (r *admissionReviewRequest) allowWithWarning(warning string) *v1.AdmissionResponse {
+	rsp := v1.AdmissionResponse{
+		Allowed: true,
+		UID:     r.req.Request.UID,
+	}
+	rsp.Warnings = append(rsp.Warnings, warning)
+	return &rsp
+}
+
+// allow constructs an allowed AdmissionResponse
+func (r *admissionReviewRequest) allow() *v1.AdmissionResponse {
+	rsp := v1.AdmissionResponse{
+		Allowed: true,
+		UID:     r.req.Request.UID,
+	}
+	return &rsp
+}
+
+// PodEviction is the entry point / handler for the pod-eviction endpoint request.
+// Note - we do not perform anything different for dry-run requests (ar.Request.DryRun) as we do not modify any state in this webhook.
+func PodEviction(ctx context.Context, l log.Logger, ar v1.AdmissionReview, kubeClient kubernetes.Interface, pdbCache *zpdb.ZpdbCache, podEvictionCache *zpdb.PodEvictionCache) *v1.AdmissionResponse {
+	logger, _ := spanlogger.New(ctx, l, "admission.PodEviction()", tenantResolver)
+	defer logger.Finish()
+
+	var pod *corev1.Pod
+	var sts *appsv1.StatefulSet
+	var err error
+
+	request := admissionReviewRequest{req: ar, log: logger, client: k8sClient{ctx: ctx, kubeClient: kubeClient}}
+
+	// set up key=value pairs we include on all logs within this context
+	request.initLogger()
+
+	level.Info(logger).Log("msg", "considering pod eviction")
+
+	// validate that this is a valid pod eviction create request
+	if err = request.validate(); err != nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - not a valid create pod eviction request", logAllow), "err", err)
+		return request.allowWithWarning(err.Error())
+	}
+
+	// get the pod which has been requested for eviction
+	if pod, err = request.client.podByName(request.req.Request.Namespace, request.req.Request.Name); err != nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errUnableToFindPodByName))
+		return request.allowWithWarning(err.Error())
+	} else if pod == nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errUnableToFindPodByName))
+		return request.allowWithWarning(errUnableToFindPodByName)
+	}
+
+	// evicting a terminal pod should result in direct deletion of pod as it already caused disruption by the time we are evicting.
+	if canIgnorePDB(pod) {
+		level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errPodNotReady))
+		return request.allowWithWarning(errPodNotReady)
+	}
+
+	pdbConfig, err := pdbCache.Find(pod)
+	if err != nil {
+		level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - %v", logDeny, err))
+		return request.denyWithReason(err.Error(), httpStatusMisconfiguration)
+	} else if pdbConfig == nil {
+		level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logAllow, errNoPdbForPod))
+		return request.allowWithWarning(errNoPdbForPod)
```

Comment:
I don't think we need a warning in this case - there will be plenty of pods with no matching ZPDB.